doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+The two frontiers never move backward, and often move forward as the system operates.
+
+At each moment, Materialize's representation of a TVC is as a set of update triples `(data, time, diff)` and the two frontiers `since` and `upper`.
+We call this pair of set of updates and `(since, upper)` frontiers a *partial TVC* or pTVC.
+
+## Evolution
+
+Materialize maintains a (partial) map from `GlobalId`s to pTVCs.
+A `GlobalId` is never re-used to mean a different TVC.
+
+The life cycle of each `GlobalId` follows several steps:
+
+0. The `GlobalId` is initially not yet bound.
+
+    In this state, the `GlobalId` cannot be used, as we do not yet know what its initial `since` will be.
+
+1. The `GlobalId` is bound to some definition, with an initially equal `since` and `upper`.
+
+    This means that the collection is initially not useful, as it cannot be queried at any time greater or equal to `since` but not greater or equal to `upper`.
+    However, the initial value of `since` indicates that one should expect the data to eventually become available for each time greater or equal to `since`.
+    The partial TVC can now be used in queries, though results may not be immediately available.
+
+2. Repeatedly, either of the these two steps occur:
+
+    a. The `upper` frontier advances, revealing more of the future of the pTVC.
+
+    b. The `since` frontier advances, concealing some of the past of the pTVC.
+
+3. Eventually, the `since` frontier advances to the empty collection, rendering the TVC complete and closed.
+
+The `GlobalId` is never reused.
+
+# Storage and Compute
+
+Materialize produces and maintains time-varying collections in two distinct ways:
+
+*   The **Storage** layer records explicit TVCs in response to direct instruction (e.g. user DML), integrations with external sources of data, and feedback from Materialize itself.
+    The representation in the Storage layer is as explicit update triples and maintained `since` and `upper` frontiers.
+    The primary property of the Storage layer is that, even across failures, it is able to reproduce all updates for times that lie between `since` and `upper` frontiers that it has advertised.
+
+*   The  **Compute** layer uses a view definition to transform input TVCs into the exactly corresponding output TVCs.
+    The output TVC varies *exactly* with the input TVCs:
+    For each `time`, the output collection at `time` equals the view applied to the input collections at `time`.
+
+The Storage and Compute layers both track `since` and `upper` frontiers for the collections they manage.
+For both layers, the `since` frontier is controlled by their users and it does not advance unless it is allowed.
+The `upper` frontier in Storage is advanced when it has sealed a frontier and made the data up to that point durable.
+The `upper` frontier in Compute is advanced when its inputs have advanced and its computation has pushed updates fully through the view definitions.
+
+## Capabilities
```

Comment:
The `Capabilities` section talks a lot about capabilities and their interaction with the system:

> [...] system-assigned unique identifier.

> These actions are mediated by the system, and take effect only when the system confirms the action

Is it clear already which (sub)system will do the above? Intuitively I am currently thinking ADAPTER, but I might be wrong.

Reply:
It was clear to me, but not to you! :D  The intent is that each layer manages the capabilities for the ids that they bind, so STORAGE manages the "sources" capabilites, COMPUTE manages the "index" and "sink" capabilities, and ADAPTER probably manages none of them? At, least not directly. It *manipulates* them of course.

Reply:
Makes sense.